druid/src/lens.rs

```diff
@@ -141,3 +142,218 @@ where
+    };
+}
+
+/// `Lens` composed of two lenses joined together
+pub struct Then<T, U, B: ?Sized> {
+    left: T,
+    right: U,
+    _marker: PhantomData<B>,
+}
+
+impl<T, U, B: ?Sized> Then<T, U, B> {
+    pub fn new<A: ?Sized, C: ?Sized>(left: T, right: U) -> Self
+    where
+        T: Lens<A, B>,
+        U: Lens<B, C>,
+    {
+        Self {
+            left,
+            right,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T, U, A, B, C> Lens<A, C> for Then<T, U, B>
+where
+    T: Lens<A, B>,
+    U: Lens<B, C>,
+{
+    fn with<V, F: FnOnce(&C) -> V>(&self, data: &A, f: F) -> V {
+        self.left.with(data, |b| self.right.with(b, f))
+    }
+
+    fn with_mut<V, F: FnOnce(&mut C) -> V>(&self, data: &mut A, f: F) -> V {
+        self.left.with_mut(data, |b| self.right.with_mut(b, f))
+    }
+}
+
+/// `Lens` for indexing containers
+pub struct Index<I> {
+    index: I,
+}
+
+impl<I> Index<I> {
+    pub fn new(index: I) -> Self {
+        Self { index }
+    }
+}
+
+impl<T, I> Lens<T, T::Output> for Index<I>
+where
+    T: ?Sized + ops::Index<I> + ops::IndexMut<I>,
+    I: Clone,
+{
+    fn with<V, F: FnOnce(&T::Output) -> V>(&self, data: &T, f: F) -> V {
+        f(&data[self.index.clone()])
+    }
+    fn with_mut<V, F: FnOnce(&mut T::Output) -> V>(&self, data: &mut T, f: F) -> V {
+        f(&mut data[self.index.clone()])
+    }
+}
+
+/// `Lens` for translating losslessly between two representations
+pub struct Iso<From, Into> {
+    from: From,
+    into: Into,
+}
+
+impl<From, Into> Iso<From, Into> {
+    pub fn new<A, B>(from: From, into: Into) -> Self
+    where
+        From: for<'a> Fn(&'a A) -> B,
+        Into: Fn(B) -> A,
+    {
+        Self { from, into }
+    }
+}
+
+impl<From, Into, A, B> Lens<A, B> for Iso<From, Into>
+where
+    From: for<'a> Fn(&'a A) -> B,
+    Into: Fn(B) -> A,
+{
+    fn with<V, F: FnOnce(&B) -> V>(&self, data: &A, f: F) -> V {
+        f(&(self.from)(data))
+    }
+
+    fn with_mut<V, F: FnOnce(&mut B) -> V>(&self, data: &mut A, f: F) -> V {
+        let mut temp = (self.from)(&*data);
+        let x = f(&mut temp);
+        *data = (self.into)(temp);
+        x
+    }
+}
+
+/// `Lens` for a scaled value
+pub struct Mul<T, O = T> {
+    factor: T,
+    _marker: PhantomData<fn() -> O>,
+}
+
+impl<T, O> Mul<T, O> {
+    pub fn new(factor: T) -> Self {
+        Self {
+            factor,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<T, U, O> Lens<T, O> for Mul<U, O>
+where
+    for<'a> &'a T: ops::Mul<&'a U, Output = O>,
+    for<'a> O: ops::Div<&'a U, Output = T>,
+{
+    fn with<V, F: FnOnce(&O) -> V>(&self, data: &T, f: F) -> V {
+        f(&(data * &self.factor))
+    }
+
+    fn with_mut<V, F: FnOnce(&mut O) -> V>(&self, data: &mut T, f: F) -> V {
+        let mut temp = &*data * &self.factor;
+        let x = f(&mut temp);
+        *data = temp / &self.factor;
+        x
+    }
+}
+
+/// `Lens` for an offset value
```

Comment:
These are motivated by my own real-world use, where I want a slider to control a value whose range is not 0-1. They're particularly convenient because they bake in the inverse operation, which is an error/correctness risk otherwise. I'll see if I can maybe expand `Iso` into something more convenient for this case while retaining generality.

Reply:
Ahhh right. I wonder whether longer term we want to do it this way (emphasizing the simplicity of individual components and composing them), or make the slider more configurable. But yes, that is an additional use case, and argues in favor of including them.